File: Server/db/sqliteConnect.py
from cgitb import reset
import sqlite3
from time import time
from Logger import log
from sqlite3 import Error

DATABASE = r"db/data/database.db"

def runSql(conn, sql):
    try:
        c = conn.cursor()
        c.execute(sql)
    except Error as e:
        log(f"SQLite Error: {e}")
        print(e)

def checkIfColumnExists(conn, table, column):
    sql = f"pragma table_info({table});"

    cursor = conn.cursor()
    columns = map(lambda line: line[1], cursor.execute(sql).fetchall())
    
    return column in columns

def createForecastFactors(conn):
    sql = """INSERT INTO forecastFactor (month, hour, factor)
    values (?, ?, ?)
    """

    cursor = conn.cursor()

    for month in range(12):
        for hour in range(24):
            cursor.execute(sql, [month + 1, hour, 1.0])
        

def getDatabaseConnection():
    conn = None
    try:
        conn = sqlite3.connect(DATABASE)

        createReadingsTable = """ CREATE TABLE IF NOT EXISTS readings (
                                        id integer PRIMARY KEY,
                                        grid_output real,
                                        battery_charge real,
                                        pv_input real,
                                        battery_state real,
                                        timestamp integer
                                    ); """

        createForecastTable = """ CREATE TABLE IF NOT EXISTS forecasts (
                                        timestamp integer PRIMARY KEY,
                                        forecast real
                                    )"""

        createForecastFactorTable = """ CREATE TABLE IF NOT EXISTS forecastFactor (
                                        id integer PRIMARY KEY AUTOINCREMENT,
                                        month integer,
                                        hour integer,
                                        factor real,
                                        numberOfInputs integer,
                                        UNIQUE(month, hour)
                                    )"""

        createDeviceInfoTable = """ CREATE TABLE IF NOT EXISTS deviceStatus (
                                        id integer PRIMARY KEY AUTOINCREMENT,
                                        identifier text,
                                        state integer,
                                        timestamp integer
                                    )
        """

        runSql(conn, createReadingsTable)
        runSql(conn, createForecastTable)
        runSql(conn, createForecastFactorTable)
        runSql(conn, createDeviceInfoTable)

        if not checkIfColumnExists(conn, "readings", "acc_grid_output"):
            addAccGridOutputSql = "alter table readings add acc_grid_output real not null default(0.0);"
            runSql(conn, addAccGridOutputSql)

        if not checkIfColumnExists(conn, "readings", "acc_grid_input"):
            addAccGridInputSql = "alter table readings add acc_grid_input real not null default(0.0);"
            runSql(conn, addAccGridInputSql)
        
        if not checkIfColumnExists(conn, "deviceStatus", "temperature_c"):
            addAccGridInputSql = "alter table deviceStatus add temperature_c real;"
            runSql(conn, addAccGridInputSql)
        
        if not checkIfColumnExists(conn, "deviceStatus", "temperature_f"):
            addAccGridInputSql = "alter table deviceStatus add temperature_f real;"
            runSql(conn, addAccGridInputSql)
        
        if not checkIfColumnExists(conn, "deviceStatus", "temperature_f"):
            addAccGridInputSql = "alter table deviceStatus add temperature_f real;"
            runSql(conn, addAccGridInputSql)
        
        if not checkIfColumnExists(conn, "deviceStatus", "humidity"):
            addAccGridInputSql = "alter table deviceStatus add humidity real;"
            runSql(conn, addAccGridInputSql)

    except Error as e:
        log(f"SQLite Error: {e}")
        print(e)

    try:
        createForecastFactors(conn)
    except:
        pass

    return conn

def insertReading(reading, forecasts):
    sql = '''   INSERT INTO readings(grid_output, battery_charge, pv_input, battery_state, timestamp, acc_grid_output, acc_grid_input) 
                VALUES(?, ?, ?, ?, ?, ?, ?) '''

    conn = getDatabaseConnection()
    cur = conn.cursor()
    cur.execute(sql, [reading["gridOutput"], reading["batteryCharge"], reading["pvInput"], reading["batteryState"], reading["timestamp"], reading["accGridOutput"], reading["accGridInput"]])

    if forecasts is not None:
        for timestamp, forecast in forecasts.items():
            forecastUpdateSql = "insert or replace into forecasts (timestamp, forecast) values (?, ?);"
            cur.execute(forecastUpdateSql, [timestamp, forecast])

            if timestamp < time() - 21600 and forecast > 0:
                factorUpdateSQL = '''      
                                    UPDATE
                                        forecastFactor
                                    SET
                                        factor = (factor * numberOfInputs + (
                                                SELECT
                                                    avg(pv_input)
                                                FROM
                                                    readings
                                                WHERE
                                                    "timestamp" BETWEEN ? - 3600 AND ?) / ?) / (numberOfInputs + 1), numberOfInputs = numberOfInputs + 1
                                        WHERE
                                            "month" = strftime ("%m", DATETIME (?, "unixepoch"))
                                            AND "hour" = strftime ("%H", DATETIME (?, "unixepoch"));
                                            
                                '''
                
                cur.execute(factorUpdateSQL, [timestamp, timestamp, forecast, timestamp, timestamp])

    conn.commit()

    conn.close()

def getCurrentDeviceStates():
    sql = 'select * from (select * from deviceStatus order by "timestamp" desc) group by identifier;'
    conn = getDatabaseConnection()
    cur = conn.cursor()
    cur.execute(sql)

    rows = cur.fetchall()

    result = []

    for row in rows:
        result.append(
            {
                "identifier": row[1],
                "state": row[2],
                "timestamp": row[3],
                "temperature_c": row[4],
                "temperature_f": row[5],
                "humidity": row[6]
            }
        )

    conn.close()

    return result


def getCurrentPVStateMovingAverage():
    sql = 'select avg(grid_output), avg(battery_charge) from (select * from readings order by "timestamp" desc limit 20)'

    conn = getDatabaseConnection()
    cur = conn.cursor()
    cur.execute(sql)

    row = cur.fetchone()

    conn.close()

    return {
            "gridOutput": row[0],
            "batteryCharge": row[1]
        }


def getCurrentDeviceStatus(identifier):
    sql = 'select * from deviceStatus where identifier == ? order by "timestamp" desc limit 1;'

    conn = getDatabaseConnection()
    cur = conn.cursor()
    cur.execute(sql, [identifier])

    row = cur.fetchone()

    conn.close()

    return {
            "identifier": row[1],
            "state": row[2],
            "timestamp": row[3],
            "temperature_c": row[4],
            "temperature_f": row[5],
            "humidity": row[6]
        }


def saveCurrentDeviceStatus(identifier, state=None, temperature_c=None, temperature_f=None, humidity=None):
    sql = 'insert into deviceStatus(identifier, state, timestamp, temperature_c, temperature_f, humidity) values(?, ?, ?, ?, ?, ?)'

    conn = getDatabaseConnection()

    cur = conn.cursor()
    cur.execute(sql, [identifier, state, int(time()), temperature_c, temperature_f, humidity])

    conn.commit()
    conn.close()

File: Server/db/sqliteConnect.php
<?php
    class MyDB extends SQLite3
    {
        
        function __construct() {
            $this->open('db/data/database.db');
        }

        function get_current_state() {
            $result = $this->query('SELECT * from readings ORDER BY timestamp DESC LIMIT 1')->fetchArray();

            return $this->parse($result);
        }

        function get_forecasts($backwardsSeconds, $forwardsSeconds) {
            $begin = time() - $backwardsSeconds;
            $end = time() + $forwardsSeconds;

            $fetchedResult = [];
            $dataStatement = $this->prepare("
            SELECT a.'timestamp' - 3600, a.forecast * c.factor, avg(b.pv_input), avg(b.battery_charge) 
            from forecasts a 
                left join readings b on b.'timestamp' between a.'timestamp' - 7200 and a.'timestamp' - 3600 
                left join forecastFactor c on c.'month' = strftime('%m', DATETIME(a.'timestamp', 'unixepoch')) and c.'hour' = strftime('%H', DATETIME(a.'timestamp', 'unixepoch'))
            where a.'timestamp' between :left and :right 
            group by a.'timestamp' 
            order by a.'timestamp' ASC ;
            ");
            $dataStatement->bindValue(':left', $begin);
            $dataStatement->bindValue(':right', $end);
            $dataResult = $dataStatement->execute();
            
            $result = [];

            while ($element = $dataResult->fetchArray()) {

                $result[] = $this->parseForecast($element);
            }
            
            return $result;
        }

        function get_daily_history() {
            $h24 = time() - 86400;

            $fetchedResult = [];

            $dataStatement = $this->prepare("SELECT avg(id), avg(grid_output), avg(battery_charge), avg(pv_input), round(avg(battery_state)), avg(timestamp) FROM readings WHERE timestamp BETWEEN :left and :right GROUP BY timestamp / 60 / 20  ORDER BY timestamp ASC;");
            $dataStatement->bindValue(':left', $h24);
            $dataStatement->bindValue(':right', time() + 3600);
            $dataResult = $dataStatement->execute();

            $result = [];

            while ($element = $dataResult->fetchArray()) {
                $result[] = $this->parse($element);
            }
            
            return $result;

        }

        function get_next_hours_forecast($number) {
            $current = time();
            $timespan = $number * 3600;

            $currentState = $this->get_current_state();
            $forecasts = $this->get_forecasts(3600, $timespan);

            $secondsIntoHour = $current - $forecasts[0]["timestamp"];
            $partOfHour = $secondsIntoHour / 3600;

            // $result = [
            //     [
            //         "consumption" => 1,
            //         "timestamp" => time(),
            //         "maxValue" => 2,
            //         "excess" => 1,
            //         "state" => 1
            //     ],
            //     [
            //         "consumption" => 1,
            //         "timestamp" => time() + 3600,
            //         "maxValue" => 1.5,
            //         "excess" => 0.5,
            //         "state" => 0
            //     ],
            //     [
            //         "consumption" => 1,
            //         "timestamp" => time() + 3600 + 3600,
            //         "maxValue" => 2.5,
            //         "excess" => 1.5,
            //         "state" => 1
            //     ],
            //     [
            //         "consumption" => 1,
            //         "timestamp" => time() + 3600 + 3600 + 3600,
            //         "maxValue" => 3.5,
            //         "excess" => 2.5,
            //         "state" => 2
            //     ],
            //     [
            //         "consumption" => 1,
            //         "timestamp" => time() + 3600 + 3600 + 3600 + 3600,
            //         "maxValue" => 3.5,
            //         "excess" => 2.5,
            //         "state" => 2
            //     ],
            //     [
            //         "consumption" => 1,
            //         "timestamp" => time() + 3600 + 3600 + 3600 + 3600 + 3600,
            //         "maxValue" => 1.5,
            //         "excess" => 0.5,
            //         "state" => 1
            //     ]
            // ];

            $result = [];

            for ($step = 0; $step < $number; $step++) {
                $forecast = $forecasts[$step];
                $consumption = $currentState["consumption"];
                
                // THIS PART SHOULD BE CHANGED WHEN PV IS NOT LOWERED!!!
                
                $maxValue = $step == 0 ?
                    $currentState["pvSystemOutput"] :
                    $forecast["forecast"] + ($forecasts[$step + 1]["forecast"] - $forecast["forecast"]) * $partOfHour;

                // $maxValue = max($forecast["forecast"] + ($forecasts[$step + 1]["forecast"] - $forecast["forecast"]) * $partOfHour, $currentState["pvInput"]);
                
                $excess = max($maxValue - $consumption, 0);

                $state = 0;
                if ($excess > 2) {
                    $state = 2;
                } else if ($excess > 1) {
                    $state = 1;
                }

                $percent = min($excess, 2) / 2;

                $result[] = [
                    "consumption" => $consumption,
                    "timestamp" => $forecast["timestamp"] + $secondsIntoHour,
                    "maxValue" => $maxValue,
                    "excess" => $excess,
                    "state" => $state,
                    "percent" => $percent
                ];
            }

            return $result;

        }

        private function parse($data) {
            return [
                "gridOutput" => $data[1],
                "batteryCharge" => $data[2],
                "pvInput" => $data[3] + max($data[2], 0),
                "batteryState" => $data[4],
                "consumption" => $data[3] - min($data[2], 0) - $data[1],
                "pvSystemOutput" => $data[3] - min($data[2], 0),
                "timestamp" => $data[5]
            ];
        }

        private function parseForecast($data) {
            return [
                "timestamp" => $data[0],
                "forecast" => $data[1],
                "data" => $data[2] ? max($data[2] + max($data[3], 0), 0) : NULL
            ];
        }
    }
?>